File: app/src/main/java/org/antrack/app/ui/fragments/SettingsFragmentNested.java
package org.antrack.app.ui.fragments;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceFragment;

import org.antrack.app.C;
import org.antrack.app.Settings;
import org.antrack.app.libs.L;
import org.antrack.app.service.MainService;

import app.R;

public class SettingsFragmentNested extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {
    final String TAG = "SettingsFragment";

    Context context;
    Intent serviceIntent;
    Settings settings;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        settings = Settings.getInstance();
        addPreferencesFromResource(R.xml.preferences);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        context = getActivity().getApplicationContext();
        boolean enabled;
        switch (key) {
            case "enable_service":
                enabled = sharedPreferences.getBoolean(key, true);
                if (enabled) {
                    settings.put(C.S_ENABLE_SERVICE, "true");
                    serviceIntent = new Intent(context, MainService.class);
                    context.startService(serviceIntent);
                    L.d(TAG, "Service started");
                } else {
                    settings.put(C.S_ENABLE_SERVICE, "false");
                    context.stopService(serviceIntent);
                    L.d(TAG, "Service stopped");
                }
                break;
            case "start_at_boot":
                enabled = sharedPreferences.getBoolean(key, true);
                if (enabled) {
                    settings.put(C.S_START_AT_BOOT, "true");
                } else {
                    settings.put(C.S_START_AT_BOOT, "false");
                }
                break;
            case "enable_ctl":
                enabled = sharedPreferences.getBoolean(key, false);
                if (enabled) {
                    settings.put(C.S_ENABLE_CTL, "true");
                    serviceIntent = new Intent(context, MainService.class);
                    serviceIntent.setAction(C.ACTION_CTL_ENABLED);
                } else {
                    settings.put(C.S_ENABLE_CTL, "false");
                    serviceIntent = new Intent(context, MainService.class);
                    serviceIntent.setAction(C.ACTION_CTL_DISABLED);
                }
                context.startService(serviceIntent);
                break;
            case "update_interval":
                String interval = sharedPreferences.getString(key, "30");
                settings.put(C.S_UPDATE_INTERVAL, interval);
                String serviceEnabled = settings.get(C.S_ENABLE_SERVICE);
                if (serviceEnabled == null || serviceEnabled.equals(C.TRUE)) {
                    serviceIntent = new Intent(context, MainService.class);
                    context.stopService(serviceIntent);
                    context.startService(serviceIntent);
                }
                break;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);


    }

    @Override
    public void onPause() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }
}

File: app/src/main/java/org/antrack/app/ui/fragments/SettingsFragmentNested.kt
@file:Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")

package org.antrack.app.ui.fragments

import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceFragment
import app.BuildConfig
import app.R
import org.antrack.app.DEFAULT_UPDATE_INTERVAL
import org.antrack.app.Settings
import org.antrack.app.WIZARD_LAUNCH_CODE
import org.antrack.app.service.CloudService
import org.antrack.app.tests.TestRunner
import org.antrack.app.ui.WizardActivity

class SettingsFragmentNested :
    PreferenceFragment(),
    SharedPreferences.OnSharedPreferenceChangeListener {

    @Deprecated("Deprecated in Java")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addPreferencesFromResource(R.xml.preferences)

        findPreference("run_setup_wizard")?.setOnPreferenceClickListener {
            WizardActivity.launch(activity, WIZARD_LAUNCH_CODE)
            true
        }

        if (BuildConfig.DEBUG) {
            setDebugOptions()
        } else {
            removeDebugOptions()
        }
    }

    override fun onSharedPreferenceChanged(sp: SharedPreferences, key: String) {
        when (key) {
            "enable_service" -> toggleEnableService(sp, key)
            "start_at_boot" -> toggleStartAtBoot(sp, key)
            "update_interval" -> toggleUpdateInterval(sp, key)
        }
    }

    private fun toggleUpdateInterval(
        sp: SharedPreferences,
        key: String
    ) {
        val interval = sp.getLong(key, DEFAULT_UPDATE_INTERVAL)
        Settings.updateInterval = interval

        if (Settings.isServiceEnabled) {
            CloudService.stop(activity)
            CloudService.start(activity)
        }
    }

    private fun toggleStartAtBoot(
        sp: SharedPreferences,
        key: String
    ) {
        Settings.startAtBoot = sp.getBoolean(key, true)
    }

    private fun toggleEnableService(
        sp: SharedPreferences,
        key: String
    ) {
        val enabled = sp.getBoolean(key, true)
        Settings.isServiceEnabled = enabled

        if (enabled) {
            CloudService.start(activity)
        } else {
            CloudService.stop(activity)
        }
    }

    private fun setDebugOptions() {
        findPreference("run_cmd_tests")?.setOnPreferenceClickListener {
            TestRunner(activity).runCmdTests()
            true
        }

        findPreference("run_mod_tests")?.setOnPreferenceClickListener {
            TestRunner(activity).runModTests()
            true
        }

        findPreference("run_cloud_tests")?.setOnPreferenceClickListener {
            TestRunner(activity).runCloudTests()
            true
        }
    }

    private fun removeDebugOptions() {
        findPreference("testing")?.apply {
            preferenceScreen.removePreference(this)
        }

        findPreference("run_cmd_tests")?.apply {
            preferenceScreen.removePreference(this)
        }

        findPreference("run_mod_tests")?.apply {
            preferenceScreen.removePreference(this)
        }

        findPreference("run_cloud_tests")?.apply {
            preferenceScreen.removePreference(this)
        }
    }

    override fun onResume() {
        super.onResume()
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        super.onPause()
    }
}
